src/Conversion/ONNXToKrnl/Tensor/Gather.cpp

```diff
@@ -41,18 +41,28 @@ struct ONNXGatherOpLowering : public ConversionPattern {
     Value alloc = insertAllocAndDeallocSimple(
         rewriter, op, outputMemRefType, loc, shapeHelper.dimsForOutput());
 
-    // Save axis and rank info.
+    // Operands and attributes.
+    Value data = operandAdaptor.data();
+    Value indices = operandAdaptor.indices();
     int64_t axisLit = gatherOp.axis();
-    int64_t dataRank = shapeHelper.dataDims.size();
-    int64_t indicesRank = shapeHelper.indicesDims.size();
-    int64_t outputRank = shapeHelper.dimsForOutput().size();
+    int64_t dataRank = data.getType().cast<ShapedType>().getRank();
```

Comment:
`ShapedType` can be unranked, e.g. `TensorType<*xtype>`, and `getRank()` would fail. Maybe, `MemRefType` is better.

Reply:
Good point!  Using MemrefType.